Subscriptions API CLI documentation (#878)
planetlabs/planet-client-python: docs/cli/cli-subscriptions.md

## Introduction

The `planet subscriptions` command enables interaction with the 

> [Subscriptions API](https://developers.planet.com/apis/subscriptions/),

Reviewer: ```suggestion
[Subscriptions API](https://developers.planet.com/apis/subscriptions/)
```

> which enables you to set up a recurring search criteria to automatically process

Reviewer: ```suggestion
that make it possible to set up a recurring search criteria. Using `planet subscriptions`, you can automatically process
```

and deliver new imagery to a cloud bucket. It also has a powerful 'backfill' capability
to bulk order historical imagery to your area of interest. This tutorial takes you
through the main commands available in the CLI.

## Core Workflows

### Create a Subscription

> Since there is no UI to easily create subscriptions we'll start with making a new one. 

Reviewer: ```suggestion
Because currently there is no UI to easily create subscriptions, we'll start with making a new one using the CLI. 
```

              35.77102915686019
            ],
            [
              139.5648193359375,
              35.42374884923695
            ]
          ]
        ],
        "type": "Polygon"
      },
      "item_types": [
        "PSScene"
      ],
      "start_time": "2023-03-17T04:08:00.0Z"
    }
  },
  "tools": [
    {
      "type": "clip",
      "parameters": {
        "aoi": {
          "coordinates": [
            [
              [
                139.5648193359375,
                35.42374884923695
              ],
              [
                140.1031494140625,
                35.42374884923695
              ],
              [
                140.1031494140625,
                35.77102915686019
              ],
              [
                139.5648193359375,
                35.77102915686019
              ],
              [
                139.5648193359375,
                35.42374884923695
              ]
            ]
          ],
          "type": "Polygon"
        }
      }
    }
  ],
  "delivery": {
    "type": "google_cloud_storage",
    "parameters": {
      "bucket": "pl-sub-bucket",
      "credentials": "<REDACTED>"
    }
  }
}
```

> This is a full subscriptions JSON request, with the credentials redacted, so you'll have 

Reviewer: ```suggestion
This is a full Subscriptions JSON request, with the credentials redacted, so you'll have 
```

> to replace your own for it to work. Below we'll show the convenience methods that will 

Reviewer: ```suggestion
to replace your own for it to work. Below you'll find convenience methods that 
```

> help create a custom one more easily, but if you'd just like to get things working then 

Reviewer: ```suggestion
help create a custom one more easily. If you'd like to get things working for now then 
```

just replace the 'delivery' section with your cloud credentials, see the 
[core subscriptions delivery docs](https://developers.planet.com/docs/subscriptions/delivery/) 
for more information. 

> To create a new subscription with the CLI you just use the `create` command:

Reviewer: ```suggestion
To create a new subscription with the CLI, use the `create` command:
```

```
planet orders subscriptions create my-subscription.json
```

The above command assumes that you've saved the subscriptions JSON as `my-subscription.json`.

### List Subscriptions

Now that you've got a subscription working you can make use of the other commands.

```
planet subscriptions list
```

> Will output the JSON for your first 100 subscriptions. If you'd like more you can set the `--limit` 

Reviewer: ```suggestion
outputs the JSON for your first 100 subscriptions. If you'd like more you can set the `--limit` 
```

parameter higher, or you can set it to 0 and there will be no limit. 

You can get nicer formatting with `--pretty` or pipe it into `jq`, just like the other Planet
CLI's.

The `list` command also supports filtering by the status of the subscription:

```
planet subscriptions list --status running
```

> will give you just the currently active subscriptions. The other available status's are:

Reviewer: ```suggestion
gives you just the currently active subscriptions. The other available statuses are:
```

`cancelled`, `preparing`, `pending`, `completed`, `suspended`, and `failed`.

> ### Describe Subscription

Author: Note this section should change with #889 - just change the three places it says 'describe' with 'get' once that change is made.

Reviewer 2: FYI: PR #900 has been merged to `main`, which closed this ticket in reference.

To get the full details on a single subscription you can take the id from your list and use the
`describe` command:

```
planet subscriptions describe cb817760-1f07-4ee7-bba6-bcac5346343f
```

### Subscription Results

To see what items have been delivered to your cloud bucket you can use the `results` command:

```
planet subscriptions results cb817760-1f07-4ee7-bba6-bcac5346343f
```

> By default it will show the first 100 results, but as with other commands, you can use the `--limit` param to 

Reviewer: ```suggestion
By default this displays the first 100 results. As with other commands, you can use the `--limit` param to 
```

its geometry. To do this you must submit the full subscription creation JSON, so the easiest way is to
get it with `describe` and then alter the values.

```
planet subscriptions update cb817760-1f07-4ee7-bba6-bcac5346343f my-updated-subscriptions.json
```

### Cancelling Subscription

Cancelling a subscription is simple with the CLI:

```
planet subscriptions cancel cb817760-1f07-4ee7-bba6-bcac5346343f
```

That will stop the subscription from producing any more results, but it will stay in the system so you can
continue to list and describe it.

## Subscription Request Conveniences

There are a couple of commands that can assist in creating the subscription JSON, used for creation and updates.
A subscription request is a pretty complicated command, consisting of a search, a cloud delivery, as well as 
tools to process the data plus notifications on the status. 

### Catalog Request

The first place to start is the `request-catalog` command, which generates all the JSON for the 
[catalog source](https://developers.planet.com/docs/subscriptions/source/#catalog-source-type) block. The core

> of this is quite similar to a Data API search requests, though with more required fields. The minimal

Reviewer: ```suggestion
of this is quite similar to a Data API search request, though with more required fields. The minimal
```

it to match the subscription creation request at the top of this documentation page:

```
{
    "coordinates":
    [
        [
            [
                139.5648193359375,
                35.42374884923695
            ],
            [
                140.1031494140625,
                35.42374884923695
            ],
            [
                140.1031494140625,
                35.77102915686019
            ],
            [
                139.5648193359375,
                35.77102915686019
            ],
            [
                139.5648193359375,
                35.42374884923695
            ]
        ]
    ],
    "type": "Polygon"
}
```

Note this is just the coordinates of either a polygon or multipolygon - the operation
is not flexible to input like the orders command.

#### RRule

RRule lets you specify a subscription that repeats at various time intervals:

```
planet subscriptions request-catalog --item-types PSScene --asset-types ortho_analytic_8b --geometry geometry.geojson \
--start-time 2023-03-17T04:08:00.0Z --rrule 'FREQ=MONTHLY;BYMONTH=3,4,5,6,7,8,9,10'
```

> For more information on the `rrule` parameter see the [recurrance rules](https://developers.planet.com/docs/subscriptions/source/#rrules-recurrence-rules)

Reviewer: ```suggestion
For more information on the `rrule` parameter see the [recurrence rules](https://developers.planet.com/docs/subscriptions/source/#rrules-recurrence-rules)
```

documentation.  

#### Filter

You can pass in a filter from the data API:

```
planet data filter --range clear_percent gt 90 > filter.json
planet subscriptions request-catalog --item-types PSScene --asset-types ortho_analytic_8b --geometry geometry.geojson \
--start-time 2022-08-24T00:00:00-07:00 --filter filter.json
```

And you can even pipe it in directly:

```
planet data filter --range clear_percent gt 90 | planet subscriptions request-catalog --item-types PSScene --asset-types ortho_analytic_8b --geometry geometry.geojson --start-time 2022-08-24T00:00:00-07:00 --filter -
```

> Just note that you should not use geometry or date filters, as they will be ignored in favor of the `--start-time` and `--geometry` values that are required.

Reviewer: ```suggestion
Do not bother with geometry or date filters, as they will be ignored in favor of the `--start-time` and `--geometry` values that are required.
```

any pixels outside of your area of interest (99.9% of all subscriptions use the clip
tool, so it's strongly recommended to also use clip). The proper 'clip' tool for it
would be:

```
[
    {
        "type": "clip",
        "parameters": {
            "aoi": {
                "type": "Polygon",
                "coordinates": [
                    [
                        [
                            -163.828125,
                            -44.59046718130883
                        ],
                        [
                            181.7578125,
                            -44.59046718130883
                        ],
                        [
                            181.7578125,
                            78.42019327591201
                        ],
                        [
                            -163.828125,
                            78.42019327591201
                        ],
                        [
                            -163.828125,
                            -44.59046718130883
                        ]
                    ]
                ]
            }
        }
    }
]
```

> You can save this tools as `tools.json` to include in the `subscriptions request` 

Reviewer: ```suggestion
You can save these tools as `tools.json` to include in the `subscriptions request` 
```

command.

#### Additional Tools

There are some other tools that are often good to include. To use more than one tool
just put them in an array in the JSON:

The toolchain options and format are given in
[Supported Tools](https://developers.planet.com/docs/subscriptions/tools/#supported-tools)
section of the subscriptions docs:

Example: `more-tools.json`
```
[
    {
        "type": "toar",
        "parameters":
        {
            "scale_factor": 10000
        }
    },
    {
        "type": "reproject",
        "parameters":
        {
            "projection": "WGS84",
            "kernel": "cubic"
        }
    },
    {
        "type": "harmonize",
        "parameters":
        {
            "target_sensor": "Sentinel-2"
        }
    },
    {
        "type": "file_format",
        "parameters":
        {
            "format": "COG"
        }
    }
]
```

### Delivery

> One other essential block is the `delivery` JSON. Like with tools there's as of yet

Reviewer: ```suggestion
One other essential block is the `delivery` JSON. As of yet, there is no
```

> not convenience method, so you'll have to write out the JSON for this section, but we

Reviewer: ```suggestion
convenience method, so you'll have to write out the JSON for this section, but we
```

the [Subscriptions Delivery documentation](https://developers.planet.com/docs/subscriptions/delivery/).

An example of a delivery.json file that you would save as a file to pass into the 
`subscriptions request` command is:

```
{
    "type": "azure_blob_storage",
    "parameters":
    {
        "account": "accountname",
        "container": "containername",
        "sas_token": "sv=2017-04-17u0026si=writersr=cu0026sig=LGqc",
        "storage_endpoint_suffix": "core.windows.net"
    }
}
```

The main documentation page also has the parameters for Google Cloud, AWS and Oracle.

### Subscriptions Request

Once you've got all your sub-blocks of JSON saved you're ready to make a complete 
subscriptions request with the `subscriptions request` command:

```
planet subscriptions request --name 'First Subscription' --source request-catalog.json \
--tools tools.json --delivery cloud-delivery.json --pretty
```

The above will print it nicely out so you can see the full request. You can write it out
as a file, or pipe it directly into `subscriptions create` or `subscriptions update`:

```
planet subscriptions request --name 'First Subscription' --source request-catalog.json \

> --tools tools.json --delivery cloud-delivery.json | planet subscriptiosn create -

Reviewer: ```suggestion
--tools tools.json --delivery cloud-delivery.json | planet subscriptions create -
```
